File: btf/btf.c
/*
 * vock native BTF parser — zero dependencies.
 * Reads /sys/kernel/btf/vmlinux directly, builds type table.
 */
#include "btf.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>

/* On-disk BTF header */
struct btf_header {
	uint16_t magic;
	uint8_t version;
	uint8_t flags;
	uint32_t hdr_len;
	uint32_t type_off;
	uint32_t type_len;
	uint32_t str_off;
	uint32_t str_len;
};

/* On-disk btf_type (12 bytes base) */
struct btf_type_raw {
	uint32_t name_off;
	uint32_t info;    /* kind=bits[28:24], vlen=bits[15:0] */
	union {
		uint32_t size;
		uint32_t type;
	};
};

/* On-disk member (12 bytes) */
struct btf_member_raw {
	uint32_t name_off;
	uint32_t type;
	uint32_t offset; /* bit offset (or bitfield encoding if kflag) */
};

/* On-disk array info */
struct btf_array_raw {
	uint32_t type;
	uint32_t index_type;
	uint32_t nelems;
};

/* On-disk enum value */
struct btf_enum_raw {
	uint32_t name_off;
	int32_t val;
};

/* On-disk enum64 value */
struct btf_enum64_raw {
	uint32_t name_off;
	uint32_t val_lo32;
	uint32_t val_hi32;
};

#define BTF_MAGIC 0xEB9F
#define BTF_INFO_KIND(info) (((info) >> 24) & 0x1f)
#define BTF_INFO_VLEN(info) ((info) & 0xffff)
#define BTF_INFO_KFLAG(info) ((info) >> 31)

struct vock_btf {
	struct vock_btf_type *types; /* indexed by type_id (0 = void) */
	uint32_t ntypes;
	char *str_section;
	uint32_t str_len;
	void *raw;  /* mmap'd or malloc'd file contents */
	size_t raw_len;
};

static const char *btf_str(struct vock_btf *btf, uint32_t off)
{
	if (off >= btf->str_len) return "";
	return btf->str_section + off;
}

struct vock_btf *vock_btf_open(const char *path)
{
	FILE *f = fopen(path, "rb");
	if (!f) return NULL;

	fseek(f, 0, SEEK_END);
	long fsize = ftell(f);
	fseek(f, 0, SEEK_SET);
	if (fsize < (long)sizeof(struct btf_header)) { fclose(f); return NULL; }

	void *raw = malloc(fsize);
	if (!raw) { fclose(f); return NULL; }
	if (fread(raw, 1, fsize, f) != (size_t)fsize) { free(raw); fclose(f); return NULL; }
	fclose(f);

	struct btf_header *hdr = (struct btf_header *)raw;
	if (hdr->magic != BTF_MAGIC) { free(raw); return NULL; }

	struct vock_btf *btf = calloc(1, sizeof(*btf));
	btf->raw = raw;
	btf->raw_len = fsize;

	uint8_t *base = (uint8_t *)raw + hdr->hdr_len;
	uint8_t *type_sec = base + hdr->type_off;
	uint32_t type_len = hdr->type_len;
	btf->str_section = (char *)(base + hdr->str_off);
	btf->str_len = hdr->str_len;

	/* First pass: count types */
	uint32_t count = 0;
	uint8_t *p = type_sec;
	uint8_t *end = type_sec + type_len;
	while (p < end) {
		struct btf_type_raw *t = (struct btf_type_raw *)p;
		uint16_t kind = BTF_INFO_KIND(t->info);
		uint16_t vlen = BTF_INFO_VLEN(t->info);
		p += sizeof(struct btf_type_raw);
		switch (kind) {
		case BTF_KIND_INT: p += 4; break;
		case BTF_KIND_ARRAY: p += sizeof(struct btf_array_raw); break;
		case BTF_KIND_STRUCT: case BTF_KIND_UNION:
			p += vlen * sizeof(struct btf_member_raw); break;
		case BTF_KIND_ENUM: p += vlen * sizeof(struct btf_enum_raw); break;
		case BTF_KIND_ENUM64: p += vlen * sizeof(struct btf_enum64_raw); break;
		case BTF_KIND_FUNC_PROTO: p += vlen * 8; break; /* params: name_off + type */
		case BTF_KIND_DATASEC: p += vlen * 12; break;
		default: break;
		}
		count++;
	}

	/* Allocate type array (id 0 = void, ids 1..count) */
	btf->ntypes = count + 1;
	btf->types = calloc(btf->ntypes, sizeof(struct vock_btf_type));
	btf->types[0].kind = 0; /* void */

	/* Second pass: populate types */
	p = type_sec;
	uint32_t id = 1;
	while (p < end && id < btf->ntypes) {
		struct btf_type_raw *t = (struct btf_type_raw *)p;
		uint16_t kind = BTF_INFO_KIND(t->info);
		uint16_t vlen = BTF_INFO_VLEN(t->info);
		int kflag = BTF_INFO_KFLAG(t->info);
		p += sizeof(struct btf_type_raw);

		struct vock_btf_type *out = &btf->types[id];
		out->id = id;
		out->kind = kind;
		out->name = btf_str(btf, t->name_off);

		switch (kind) {
		case BTF_KIND_INT: {
			out->size = t->size;
			uint32_t enc = *(uint32_t *)p;
			out->int_bits = enc & 0xff;
			out->int_signed = (enc >> 24) & 1;
			p += 4;
			break;
		}
		case BTF_KIND_PTR: case BTF_KIND_TYPEDEF:
		case BTF_KIND_VOLATILE: case BTF_KIND_CONST:
		case BTF_KIND_RESTRICT: case BTF_KIND_FWD:
			out->ref_type_id = t->type;
			break;
		case BTF_KIND_ARRAY: {
			struct btf_array_raw *a = (struct btf_array_raw *)p;
			out->array_type_id = a->type;
			out->array_nelems = a->nelems;
			p += sizeof(struct btf_array_raw);
			break;
		}
		case BTF_KIND_STRUCT: case BTF_KIND_UNION: {
			out->size = t->size;
			out->nmembers = vlen;
			out->members = calloc(vlen, sizeof(struct vock_btf_member));
			struct btf_member_raw *m = (struct btf_member_raw *)p;
			for (int i = 0; i < vlen; i++) {
				out->members[i].name = btf_str(btf, m[i].name_off);
				out->members[i].type_id = m[i].type;
				if (kflag)
					out->members[i].offset_bits = m[i].offset & 0xffffff;
				else
					out->members[i].offset_bits = m[i].offset;
			}
			p += vlen * sizeof(struct btf_member_raw);
			break;
		}
		case BTF_KIND_ENUM: {
			out->size = t->size;
			out->nenums = vlen;
			out->enums = calloc(vlen, sizeof(struct vock_btf_enum_val));
			struct btf_enum_raw *e = (struct btf_enum_raw *)p;
			for (int i = 0; i < vlen; i++) {
				out->enums[i].name = btf_str(btf, e[i].name_off);
				out->enums[i].val = e[i].val;
			}
			p += vlen * sizeof(struct btf_enum_raw);
			break;
		}
		case BTF_KIND_ENUM64: {
			out->size = t->size;
			out->nenums = vlen;
			out->enums = calloc(vlen, sizeof(struct vock_btf_enum_val));
			struct btf_enum64_raw *e = (struct btf_enum64_raw *)p;
			for (int i = 0; i < vlen; i++) {
				out->enums[i].name = btf_str(btf, e[i].name_off);
				out->enums[i].val = (int64_t)((uint64_t)e[i].val_hi32 << 32 | e[i].val_lo32);
			}
			p += vlen * sizeof(struct btf_enum64_raw);
			break;
		}
		case BTF_KIND_FUNC_PROTO:
			out->ref_type_id = t->type; /* return type */
			p += vlen * 8;
			break;
		case BTF_KIND_FUNC:
			out->ref_type_id = t->type;
			break;
		case BTF_KIND_VAR:
			out->ref_type_id = t->type;
			p += 4; /* linkage */
			break;
		case BTF_KIND_DATASEC:
			out->size = t->size;
			p += vlen * 12;
			break;
		case BTF_KIND_FLOAT:
			out->size = t->size;
			break;
		default:
			break;
		}
		id++;
	}

	return btf;
}

void vock_btf_close(struct vock_btf *btf)
{
	if (!btf) return;
	for (uint32_t i = 0; i < btf->ntypes; i++) {
		free(btf->types[i].members);
		free(btf->types[i].enums);
	}
	free(btf->types);
	free(btf->raw);
	free(btf);
}

uint32_t vock_btf_type_count(struct vock_btf *btf)
{
	return btf->ntypes;
}

const struct vock_btf_type *vock_btf_type_by_id(struct vock_btf *btf, uint32_t id)
{
	if (id >= btf->ntypes) return NULL;
	return &btf->types[id];
}

const struct vock_btf_type *vock_btf_find_struct(struct vock_btf *btf, const char *name)
{
	for (uint32_t i = 1; i < btf->ntypes; i++) {
		if ((btf->types[i].kind == BTF_KIND_STRUCT ||
		     btf->types[i].kind == BTF_KIND_UNION) &&
		    btf->types[i].name && strcmp(btf->types[i].name, name) == 0)
			return &btf->types[i];
	}
	return NULL;
}

static const char *kind_str(uint16_t kind)
{
	switch (kind) {
	case BTF_KIND_INT: return "int";
	case BTF_KIND_PTR: return "ptr";
	case BTF_KIND_ARRAY: return "array";
	case BTF_KIND_STRUCT: return "struct";
	case BTF_KIND_UNION: return "union";
	case BTF_KIND_ENUM: return "enum";
	case BTF_KIND_TYPEDEF: return "typedef";
	case BTF_KIND_CONST: return "const";
	case BTF_KIND_VOLATILE: return "volatile";
	case BTF_KIND_FLOAT: return "float";
	default: return "?";
	}
}

void vock_btf_dump_struct(struct vock_btf *btf, const struct vock_btf_type *t, int depth)
{
	if (!t || depth > 4) return;
	const char *kw = t->kind == BTF_KIND_UNION ? "union" : "struct";
	printf("%s %s { /* %u bytes */\n", kw, t->name ? t->name : "(anon)", t->size);
	for (int i = 0; i < t->nmembers; i++) {
		struct vock_btf_member *m = &t->members[i];
		const struct vock_btf_type *mt = vock_btf_type_by_id(btf, m->type_id);
		/* Chase qualifiers */
		while (mt && (mt->kind == BTF_KIND_CONST || mt->kind == BTF_KIND_VOLATILE ||
		             mt->kind == BTF_KIND_TYPEDEF || mt->kind == BTF_KIND_RESTRICT))
			mt = vock_btf_type_by_id(btf, mt->ref_type_id);
		printf("  +%3u.%u  %-20s  ", m->offset_bits / 8, m->offset_bits % 8,
		       m->name[0] ? m->name : "(anon)");
		if (!mt) { printf("void\n"); continue; }
		if (mt->kind == BTF_KIND_INT)
			printf("%s%u\n", mt->int_signed ? "s" : "u", mt->int_bits);
		else if (mt->kind == BTF_KIND_PTR)
			printf("ptr\n");
		else if (mt->kind == BTF_KIND_ENUM)
			printf("enum %s (%d vals)\n", mt->name ? mt->name : "", mt->nenums);
		else if (mt->kind == BTF_KIND_STRUCT || mt->kind == BTF_KIND_UNION)
			printf("%s %s (%u bytes)\n", kind_str(mt->kind),
			       mt->name ? mt->name : "(anon)", mt->size);
		else if (mt->kind == BTF_KIND_ARRAY)
			printf("array[%u]\n", mt->array_nelems);
		else
			printf("%s\n", kind_str(mt->kind));
	}
	printf("}\n");
}

File: vock/src/fuzz/btf.rs
//! Native BTF parser (port of btf/btf.c), zero dependencies.
//!
//! Reads `/sys/kernel/btf/vmlinux` (raw BTF) or an ELF file with a `.BTF`
//! section directly and builds an in-memory type table indexed by type id
//! (id 0 == void).
#![allow(dead_code)]

// BTF type kinds.
pub const BTF_KIND_INT: u16 = 1;
pub const BTF_KIND_PTR: u16 = 2;
pub const BTF_KIND_ARRAY: u16 = 3;
pub const BTF_KIND_STRUCT: u16 = 4;
pub const BTF_KIND_UNION: u16 = 5;
pub const BTF_KIND_ENUM: u16 = 6;
pub const BTF_KIND_FWD: u16 = 7;
pub const BTF_KIND_TYPEDEF: u16 = 8;
pub const BTF_KIND_VOLATILE: u16 = 9;
pub const BTF_KIND_CONST: u16 = 10;
pub const BTF_KIND_RESTRICT: u16 = 11;
pub const BTF_KIND_FUNC: u16 = 12;
pub const BTF_KIND_FUNC_PROTO: u16 = 13;
pub const BTF_KIND_VAR: u16 = 14;
pub const BTF_KIND_DATASEC: u16 = 15;
pub const BTF_KIND_FLOAT: u16 = 16;
pub const BTF_KIND_ENUM64: u16 = 19;

const BTF_MAGIC: u16 = 0xEB9F;

pub struct BtfMember {
    pub name: String,
    pub type_id: u32,
    pub offset_bits: u32,
}

pub struct BtfEnumVal {
    pub name: String,
    pub val: i64,
}

pub struct BtfType {
    pub id: u32,
    pub kind: u16,
    pub name: String,
    pub size: u32,        // byte size for INT/STRUCT/UNION/ENUM
    pub ref_type_id: u32, // for PTR/TYPEDEF/CONST/VOLATILE/RESTRICT
    pub members: Vec<BtfMember>,
    pub enums: Vec<BtfEnumVal>,
    pub int_bits: u8,
    pub int_signed: u8,
    pub array_type_id: u32,
    pub array_nelems: u32,
}

impl BtfType {
    fn void() -> BtfType {
        BtfType {
            id: 0,
            kind: 0,
            name: String::new(),
            size: 0,
            ref_type_id: 0,
            members: Vec::new(),
            enums: Vec::new(),
            int_bits: 0,
            int_signed: 0,
            array_type_id: 0,
            array_nelems: 0,
        }
    }
}

pub struct Btf {
    pub types: Vec<BtfType>, // index == type id; [0] == void
}

fn rd_u16(b: &[u8], off: usize) -> Option<u16> {
    b.get(off..off + 2).map(|s| u16::from_le_bytes([s[0], s[1]]))
}
fn rd_u32(b: &[u8], off: usize) -> Option<u32> {
    b.get(off..off + 4)
        .map(|s| u32::from_le_bytes([s[0], s[1], s[2], s[3]]))
}
fn rd_u64(b: &[u8], off: usize) -> Option<u64> {
    b.get(off..off + 8).map(|s| {
        u64::from_le_bytes([s[0], s[1], s[2], s[3], s[4], s[5], s[6], s[7]])
    })
}

fn btf_str(strs: &[u8], off: u32) -> String {
    let off = off as usize;
    if off >= strs.len() {
        return String::new();
    }
    let end = strs[off..]
        .iter()
        .position(|&c| c == 0)
        .map(|p| off + p)
        .unwrap_or(strs.len());
    String::from_utf8_lossy(&strs[off..end]).into_owned()
}

/// Locate the `.BTF` section inside an ELF64 image. Returns (offset, size).
fn find_elf_btf(raw: &[u8]) -> Option<(usize, usize)> {
    // Elf64_Ehdr fields we need.
    let e_shoff = rd_u64(raw, 0x28)? as usize;
    let e_shentsize = rd_u16(raw, 0x3a)? as usize;
    let e_shnum = rd_u16(raw, 0x3c)? as usize;
    let e_shstrndx = rd_u16(raw, 0x3e)? as usize;
    if e_shentsize < 64 || e_shoff + e_shnum * e_shentsize > raw.len() {
        return None;
    }
    let shdr_at = |i: usize| e_shoff + i * e_shentsize;
    // section header string table
    let strtab_off = rd_u64(raw, shdr_at(e_shstrndx) + 0x18)? as usize;
    for i in 0..e_shnum {
        let base = shdr_at(i);
        let sh_name = rd_u32(raw, base)? as usize;
        let sh_offset = rd_u64(raw, base + 0x18)? as usize;
        let sh_size = rd_u64(raw, base + 0x20)? as usize;
        let name_ptr = strtab_off + sh_name;
        let name_end = raw[name_ptr..]
            .iter()
            .position(|&c| c == 0)
            .map(|p| name_ptr + p)
            .unwrap_or(raw.len());
        if &raw[name_ptr..name_end] == b".BTF" {
            return Some((sh_offset, sh_size));
        }
    }
    None
}

impl Btf {
    pub fn open(path: &str) -> Option<Btf> {
        let raw = std::fs::read(path).ok()?;
        if raw.len() < 24 {
            return None;
        }

        // Extract the .BTF section if this is an ELF file.
        let (bstart, bsize) = if raw.len() > 64 && &raw[0..4] == b"\x7fELF" {
            find_elf_btf(&raw)?
        } else {
            (0, raw.len())
        };
        let btf_data = raw.get(bstart..bstart + bsize)?;
        if btf_data.len() < 24 || rd_u16(btf_data, 0)? != BTF_MAGIC {
            return None;
        }

        // BTF header.
        let hdr_len = rd_u32(btf_data, 4)? as usize;
        let type_off = rd_u32(btf_data, 8)? as usize;
        let type_len = rd_u32(btf_data, 12)? as usize;
        let str_off = rd_u32(btf_data, 16)? as usize;
        let str_len = rd_u32(btf_data, 20)? as usize;

        let base = hdr_len;
        let type_start = base + type_off;
        let type_end = type_start + type_len;
        let str_start = base + str_off;
        let strs = btf_data.get(str_start..str_start + str_len)?;
        let types_sec = btf_data.get(type_start..type_end)?;

        let mut types: Vec<BtfType> = Vec::new();
        types.push(BtfType::void());

        let mut p = 0usize;
        let mut id: u32 = 1;
        while p + 12 <= types_sec.len() {
            let name_off = rd_u32(types_sec, p)?;
            let info = rd_u32(types_sec, p + 4)?;
            let size_or_type = rd_u32(types_sec, p + 8)?;
            let kind = ((info >> 24) & 0x1f) as u16;
            let vlen = (info & 0xffff) as usize;
            let kflag = info >> 31;
            p += 12;

            let mut out = BtfType::void();
            out.id = id;
            out.kind = kind;
            out.name = btf_str(strs, name_off);

            match kind {
                BTF_KIND_INT => {
                    out.size = size_or_type;
                    let enc = rd_u32(types_sec, p)?;
                    out.int_bits = (enc & 0xff) as u8;
                    out.int_signed = ((enc >> 24) & 1) as u8;
                    p += 4;
                }
                BTF_KIND_PTR | BTF_KIND_TYPEDEF | BTF_KIND_VOLATILE | BTF_KIND_CONST
                | BTF_KIND_RESTRICT | BTF_KIND_FWD => {
                    out.ref_type_id = size_or_type;
                }
                BTF_KIND_ARRAY => {
                    out.array_type_id = rd_u32(types_sec, p)?;
                    out.array_nelems = rd_u32(types_sec, p + 8)?;
                    p += 12;
                }
                BTF_KIND_STRUCT | BTF_KIND_UNION => {
                    out.size = size_or_type;
                    out.members.reserve(vlen);
                    for i in 0..vlen {
                        let mp = p + i * 12;
                        let m_name = rd_u32(types_sec, mp)?;
                        let m_type = rd_u32(types_sec, mp + 4)?;
                        let m_off = rd_u32(types_sec, mp + 8)?;
                        let offset_bits = if kflag != 0 { m_off & 0xffffff } else { m_off };
                        out.members.push(BtfMember {
                            name: btf_str(strs, m_name),
                            type_id: m_type,
                            offset_bits,
                        });
                    }
                    p += vlen * 12;
                }
                BTF_KIND_ENUM => {
                    out.size = size_or_type;
                    out.enums.reserve(vlen);
                    for i in 0..vlen {
                        let ep = p + i * 8;
                        let e_name = rd_u32(types_sec, ep)?;
                        let e_val = rd_u32(types_sec, ep + 4)? as i32 as i64;
                        out.enums.push(BtfEnumVal {
                            name: btf_str(strs, e_name),
                            val: e_val,
                        });
                    }
                    p += vlen * 8;
                }
                BTF_KIND_ENUM64 => {
                    out.size = size_or_type;
                    out.enums.reserve(vlen);
                    for i in 0..vlen {
                        let ep = p + i * 12;
                        let e_name = rd_u32(types_sec, ep)?;
                        let lo = rd_u32(types_sec, ep + 4)? as u64;
                        let hi = rd_u32(types_sec, ep + 8)? as u64;
                        out.enums.push(BtfEnumVal {
                            name: btf_str(strs, e_name),
                            val: ((hi << 32) | lo) as i64,
                        });
                    }
                    p += vlen * 12;
                }
                BTF_KIND_FUNC_PROTO => {
                    out.ref_type_id = size_or_type;
                    p += vlen * 8;
                }
                BTF_KIND_FUNC => {
                    out.ref_type_id = size_or_type;
                }
                BTF_KIND_VAR => {
                    out.ref_type_id = size_or_type;
                    p += 4;
                }
                BTF_KIND_DATASEC => {
                    out.size = size_or_type;
                    p += vlen * 12;
                }
                BTF_KIND_FLOAT => {
                    out.size = size_or_type;
                }
                _ => {}
            }

            types.push(out);
            id += 1;
        }

        Some(Btf { types })
    }

    pub fn type_count(&self) -> u32 {
        self.types.len() as u32
    }

    pub fn type_by_id(&self, id: u32) -> Option<&BtfType> {
        self.types.get(id as usize)
    }

    pub fn find_struct(&self, name: &str) -> Option<&BtfType> {
        self.types.iter().skip(1).find(|t| {
            (t.kind == BTF_KIND_STRUCT || t.kind == BTF_KIND_UNION) && t.name == name
        })
    }

    /// Resolve a type through const/volatile/typedef/restrict qualifiers.
    pub fn resolve<'a>(&'a self, mut t: &'a BtfType) -> &'a BtfType {
        while matches!(
            t.kind,
            BTF_KIND_CONST | BTF_KIND_VOLATILE | BTF_KIND_TYPEDEF | BTF_KIND_RESTRICT
        ) {
            match self.type_by_id(t.ref_type_id) {
                Some(next) => t = next,
                None => break,
            }
        }
        t
    }
}
